osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs

```diff
@@ -0,0 +1,115 @@
+        /// </remarks>
+        private readonly Dictionary<HitObjectLifetimeEntry, HitObject?> parentMap = new Dictionary<HitObjectLifetimeEntry, HitObject?>();
+
+        /// <summary>
+        /// Stores the list of entries managed by this <see cref="HitObjectEntryManager"/> for each hit object managed by this <see cref="HitObjectEntryManager"/>.
+        /// </summary>
+        private readonly Dictionary<HitObject, List<HitObjectLifetimeEntry>> childrenMap = new Dictionary<HitObject, List<HitObjectLifetimeEntry>>();
+
+        public void Add(HitObjectLifetimeEntry entry, HitObject? parent)
+        {
+            if (parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is already added to this {nameof(HitObjectEntryManager)}.");
+
+            var hitObject = entry.HitObject;
+            parentMap[entry] = parent;
+            entryMap[hitObject] = entry;
+
+            if (parent != null && childrenMap.TryGetValue(parent, out var parentChildEntries))
+                parentChildEntries.Add(entry);
+
+            hitObject.DefaultsApplied += onDefaultsApplied;
+
+            childrenMap[hitObject] = new List<HitObjectLifetimeEntry>();
+
+            OnEntryAdded?.Invoke(entry, parent);
+        }
+
+        public void Remove(HitObjectLifetimeEntry entry)
+        {
+            if (!parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is not contained in this {nameof(HitObjectLifetimeEntry)}.");
+
+            var hitObject = entry.HitObject;
+            parentMap.Remove(entry, out var parent);
+            entryMap.Remove(hitObject);
+
+            if (parent != null && childrenMap.TryGetValue(parent, out var parentChildEntries))
+                parentChildEntries.Remove(entry);
+
+            hitObject.DefaultsApplied -= onDefaultsApplied;
+
+            // Remove all entries of the nested hit objects
+            if (childrenMap.Remove(entry.HitObject, out var childEntries))
+            {
+                foreach (var childEntry in childEntries)
+                    Remove(childEntry);
+            }
+
+            OnEntryRemoved?.Invoke(entry, parent);
+        }
+
+        public bool TryGet(HitObject hitObject, [MaybeNullWhen(false)] out HitObjectLifetimeEntry entry)
+        {
+            return entryMap.TryGetValue(hitObject, out entry);
+        }
+
+        /// <summary>
+        /// As nested hit objects are recreated, remove entries of the old nested hit objects.
+        /// </summary>
+        private void onDefaultsApplied(HitObject hitObject)
+        {
+            if (!childrenMap.Remove(hitObject, out var childEntries))
+                return;
+
+            foreach (var entry in childEntries)
+                Remove(entry);
+
+            childEntries.Clear();
+            childrenMap[hitObject] = childEntries;
```

Comment:
I also don't understand the purpose of this line in general. As far as I understand it, this is what this method does:
- `childrenMap[hitObject] -> childEntries`
- `childEntries.Clear()`
- `childrenMap[hitObject] <- childEntries`
It looks like this operation, the last one in my list above, is a no-op? Am I missing something?

Reply:
I understand now why this is done the way it is, everything relies on that `.Remove()` call. I'm going to refactor this because it's hard to understand and easy to miss.

Reply:
Well, I didn't end up changing this. The idea I had was to not do the whole `.Remove()` stuff altogether, but in hindsight this is probably the more performance way of doing things anyway.

I added what I believe is appropriate commenting along the way, tho.